File: ports/ts/src/fakes/fakes.ts
// Deterministic fakes for the hermetic ask flow (SPEC-PORTS-v1 §0/§4).
//
// These are the offline stand-ins that make cite-or-abstain provable without a
// model server: a hash-bucket embedder, an evidence-echoing LLM, and a cosine
// that is a plain dot product (vectors arrive L2-normalized). Parity with the
// Python reference citenexus.testing.fakes.

import { createHash } from "node:crypto";
import { tokenize } from "../tokenize/tokenize.js";

/** Embedding dimensionality (§4). */
export const EMBED_DIM = 64;

/**
 * Hash-bucket bag-of-tokens embedder. Each token lands in one of EMBED_DIM
 * buckets via sha1(token) mod EMBED_DIM; the vector is then L2-normalized (an
 * all-zero vector stays zero). Deterministic and language-agnostic.
 */
export class FakeEmbedding {
  embed(text: string): number[] {
    const vec = new Array<number>(EMBED_DIM).fill(0);
    for (const token of tokenize(text)) {
      const hex = createHash("sha1").update(token, "utf8").digest("hex");
      const idx = Number(BigInt("0x" + hex) % BigInt(EMBED_DIM));
      vec[idx] = (vec[idx] ?? 0) + 1.0;
    }
    let sumSquares = 0;
    for (const v of vec) sumSquares += v * v;
    const norm = Math.sqrt(sumSquares);
    if (norm === 0) return vec;
    return vec.map((v) => v / norm);
  }
}

/** Evidence-echoing generator: the answer is the cited passage, verbatim. */
export class FakeLLM {
  answer(_question: string, passage: string): string {
    return passage;
  }
}

/** Cosine similarity of two already-normalized vectors: a plain dot product. */
export function cosine(a: readonly number[], b: readonly number[]): number {
  let dot = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    dot += (a[i] ?? 0) * (b[i] ?? 0);
  }
  return dot;
}

File: golang/fakes/fakes.go
// Package fakes holds the deterministic, offline test doubles that make the
// CiteNexus cite-or-abstain flow provable without any model server
// (SPEC-PORTS-v1 §0/§4). They mirror the Python reference
// citenexus.testing.fakes: a hash-bucketed embedding, an evidence-echoing LLM,
// and a cosine over already-normalized vectors.
package fakes

import (
	"crypto/sha1"
	"math"

	"github.com/muthuishere/citenexus/golang/tokenize"
)

// Dim is the fixed embedding dimensionality (§4).
const Dim = 64

// FakeEmbedding is a hash-bucketed bag-of-tokens embedder: each pinned token is
// hashed into one of Dim buckets and increments it, then the vector is
// L2-normalized. Deterministic and offline — the property that makes the
// guarantee testable.
type FakeEmbedding struct{}

// Embed maps text to a unit vector of length Dim. For each token produced by the
// pinned tokenizer, bucket = sha1(utf8(token)) mod Dim (the full 160-bit digest
// mod 64 collapses to the low 6 bits of its least-significant byte). The result
// is L2-normalized; an all-zero vector (no tokens) is left as zeros.
func (FakeEmbedding) Embed(text string) []float64 {
	vec := make([]float64, Dim)
	for _, tok := range tokenize.Tokenize(text) {
		sum := sha1.Sum([]byte(tok))
		idx := int(sum[len(sum)-1] & 0x3F) // digest mod 64 == low 6 bits.
		vec[idx] += 1.0
	}
	var norm float64
	for _, v := range vec {
		norm += v * v
	}
	norm = math.Sqrt(norm)
	if norm != 0 {
		for i := range vec {
			vec[i] /= norm
		}
	}
	return vec
}

// FakeLLM is an evidence-echoing generator: it returns the cited passage
// verbatim, so the faithfulness gate is exercised honestly rather than bypassed.
type FakeLLM struct{}

// Answer returns the passage unchanged, ignoring the question.
func (FakeLLM) Answer(question, passage string) string { return passage }

// Cosine is the dot product of a and b. FakeEmbedding vectors are already
// L2-normalized, so the dot product is their cosine similarity.
func Cosine(a, b []float64) float64 {
	var dot float64
	for i := range a {
		dot += a[i] * b[i]
	}
	return dot
}
